model/src/main/kotlin/utils/DependencyGraphBuilder.kt

```diff
@@ -167,7 +168,19 @@ class DependencyGraphBuilder<D>(
 
         val (nodes, edges) = directDependencies.toGraph()
 
-        return DependencyGraph(dependencyIds, sortedSetOf(), scopeMapping, nodes, edges)
+        return DependencyGraph(dependencyIds, sortedSetOf(), scopeMapping, nodes, edges.removeAllCycles())
+    }
+
+    private fun Collection<DependencyGraphEdge>.removeAllCycles(): List<DependencyGraphEdge> {
+        val edges = mapTo(mutableSetOf()) { it.from to it.to }
```

Comment:
Does it make things easier to use `associate { it.from to it.to }` here and work on a map rather than a set of pairs?

Reply:
IMO not, because `edges` are passed to `breakAllCycles(edges: Collection<Pair<Int, Int>>)`.
Making the parameter a `Map` would mean that each node can only have one outgoing edge which is not desired.